File: vn_custom/doc_events/sales_invoice.py
# -*- coding: utf-8 -*-
from vn_custom.doc_events.delivery_note import set_item_prices
from vn_custom.api.item import get_other_prices
from vn_custom.utils import pick


def on_update(doc, method):
    for item in doc.items:
        if not item.vn_mrp or not item.vn_valuation:
            prices = get_other_prices(item.item_code, item.uom)
            item.update(pick(["vn_mrp", "vn_valuation"], prices))


def on_submit(doc, method):
    if doc.update_stock:
        set_item_prices(doc.items, doc.posting_date, doc.posting_time)


def on_cancel(doc, method):
    if doc.update_stock:
        set_item_prices(doc.items, doc.posting_date, doc.posting_time)

File: vn_custom/public/js/custom_scripts/sales_invoice.js
async function set_other_prices(frm, cdt, cdn) {
  const { item_code, uom } = frappe.get_doc(cdt, cdn) || {};
  if (item_code) {
    const {
      message: { vn_mrp = 0, vn_valuation = 0 } = {},
    } = await frappe.call({
      method: 'vn_custom.api.item.get_other_prices',
      args: { item_code, uom },
    });
    frappe.model.set_value(cdt, cdn, { vn_mrp, vn_valuation });
  } else {
    frappe.model.set_value(cdt, cdn, { vn_mrp: null, vn_valuation: null });
  }
  },
};

const sales_invoice = {
  setup: function (frm) {
    console.log('setup');
    frm.set_query('uom', 'items', function (doc, cdt, cdn) {
      const { item_code } = frappe.get_doc(cdt, cdn);
      if (item_code) {
        return {
          query: 'vn_custom.api.item.get_uom',
          filters: { item_code },
        };
}
    });
  },
};

export default {
  sales_invoice_item,
  sales_invoice,
};
